### DISCUSSION TITLE
Enable usage of KDL markup language instead of XML for mapping files.

### DISCUSSION THREAD

**Discussion by jkoritzinsky** [SharpGenTools/SharpGenTools#202] [Ideas]
[KDL](https://kdl.dev/) is a newer markup language that is node+attribute based like XML but is cleaner. Additionally, the nodes look like command line statements, which fits the pseudo-imperative model our mapping files use much better than the current XML support.

There's an experimental .NET parser for KDL at https://github.com/borland/kdl-net.

cc: @andrew-boyarshin this is the new markup language issue I was thinking of.

**Comment by andrew-boyarshin** [comment]
I'm wary about using little known format (yeah, that one xkcd comic about standards), but it's actually pretty nice replacement for XML. I'm not sure if it fits Avalonia, though. KDL is great about structure, not about markup. WPF (and Avalonia) need actual markup (`Span`, `Run`). But for our purposes it feels sufficient. There are 2 things that I feel KDL is lacking:
1. Conditionals, even if in the same form as in MSBuild. It would be cool to have that. Imagine mapping `long` type to `int` or `long` depending on the ABI (LLP64 vs LP64).
2. Language services. Nonexistent with KDL. No proper syntax highlighting. VSCode is the only editor with basic support. On the other hand, XML is supported everywhere, my favorite .NET parser is [XmlParser](https://github.com/KirillOsenkov/XmlParser).

My previous thoughts on a new format for mapping files were basically MSBuild project files. This way we'd get language services and good integration with build system (including proper dependency tracking). The downside is the XML verbosity. I also considered [GN language](https://gn.googlesource.com/gn/+/master/docs/language.md) (used to build Chromium & Fuchsia, very fast) and [Starlark](https://github.com/google/starlark-go) (used by Bazel and Buck, limited subset of Python). They are cool, but not ideal for our purposes.
